### PR TITLE
Merge CacheDirectory into SearchableSnapshotDirectory

### PR DESCRIPTION
**Pull request by tlrx**
Following #53860 and #53879, this pull request now merges `CacheDirectory` into `SearchableSnapshotDirectory`.

I expect few more follow up pull requests, mostly to clean up things and move classes.

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotDirectory.java [hunk 1]
```diff
@@ -1,109 +0,0 @@
...
-import org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
-
-import java.io.FileNotFoundException;
-import java.io.IOException;
-import java.util.Collection;
-import java.util.Objects;
-import java.util.Set;
-import java.util.concurrent.atomic.AtomicBoolean;
-
-public abstract class BaseSearchableSnapshotDirectory extends BaseDirectory {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotDirectory.java]: This class has been merged down with `SearchableSnapshotDirectory` (it was useful to me for the 2 previous PRs)

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 2]
```diff
@@ -42,17 +60,175 @@
...
+    public ShardId getShardId() {
+        return shardId;
+    }
+
+    public Map<String, IndexInputStats> getStats() {
+        return Collections.unmodifiableMap(stats);
+    }
+
+    @Nullable
+    public IndexInputStats getStats(String fileName) {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java]: Comes from the now removed CacheDirectory and made `public` as cache classes are in a different package

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 3]
```diff
@@ -42,17 +60,175 @@
...
         ensureOpen();
-        return new SearchableSnapshotIndexInput(blobContainer, fileInfo(name), context, blobContainer.readBlobPreferredLength(),
-            BufferedIndexInput.BUFFER_SIZE);
+        final BlobStoreIndexShardSnapshot.FileInfo fileInfo = fileInfo(name);
+        final IndexInputStats inputStats = stats.computeIfAbsent(name, n -> createIndexInputStats(fileInfo.length()));
+        if (useCache) {
+            return new CacheBufferedIndexInput(this, fileInfo, context, inputStats);
+        } else {
+            long preferredLength = blobContainer.readBlobPreferredLength();
+            return new SearchableSnapshotIndexInput(blobContainer, fileInfo, context, preferredLength, BufferedIndexInput.BUFFER_SIZE);
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java]: I'd like to rename this one to DirectBufferedIndexInput and to clean it up a bit in a follow up PR

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java [hunk 4]
```diff
@@ -47,13 +48,13 @@
     // last seek position is kept around in order to detect forward/backward seeks for stats
     private long lastSeekPosition;
 
-    CacheBufferedIndexInput(CacheDirectory directory, FileInfo fileInfo, IOContext context, IndexInputStats stats) {
+    public CacheBufferedIndexInput(SearchableSnapshotDirectory directory, FileInfo fileInfo, IOContext context, IndexInputStats stats) {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheBufferedIndexInput.java]: Has to be `public` now it's instantiated by `SearchableSnapshotDirectory` from a different package.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheFile.java [hunk 5]
```diff
@@ -27,7 +27,7 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 
-class CacheFile {
+public class CacheFile {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheFile.java]: Same here.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheKey.java [hunk 6]
```diff
@@ -18,7 +18,7 @@
     private final ShardId shardId;
     private final String fileName;
 
-    CacheKey(SnapshotId snapshotId, IndexId indexId, ShardId shardId, String fileName) {
+    public CacheKey(SnapshotId snapshotId, IndexId indexId, ShardId shardId, String fileName) {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheKey.java]: And same here.

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java [hunk 7]
```diff
@@ -380,10 +405,86 @@ private void testIndexInputs(final CheckedBiConsumer<IndexInput, IndexInput, Exc
         });
     }
 
+    public void testClearCache() throws Exception {
```
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotDirectoryTests.java]: This test comes from the now removed `CacheDirectoryTests` class with very few adjustments for the `SearchableSnapshotDirectory` instantiation.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks David

**Pull request merged**